### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by pug000**
1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/tasks/stage-2/rs-lang/rslang.md
2. Screenshot: 
Снимок экрана 2022-09-05 в 22 41 19

3. Deploy: https://pug000.github.io/rslang/
4. Done 05.09.2022 / deadline 06.09.2022
5.  Score (660 / 700)

- **Application main page** (+40):
  -   - [x] section navigation menu, links to mini-games (+10);
  -   - [x] section "About" with description of the features and benefits of the application (+10);
  -   - [x] section "About" with avatars and links to githubs of all team members (+10);
  -   - [x] footer with links to the githubs of the authors of the application, the year the application was created, the course logo with a link to the course. Footer is displayed on all pages of the application except mini-games (+10);

- **Authorization** (+50):
  -   - [x] user authorization and registration implemented (+10);
  -   - [x] when the application is reloaded, the user data and the token obtained during authorization are stored in localStorage (+10);
  -   - [x] electronic textbook (without the "Difficult words" section, without displaying the progress of learning words and learned words) and mini-games are available without authorization (+10);
  -   - [x] authorized users see the "Difficult words" section, word learning progress, learned words and statistics (+10);
  -   - [x] upon logout, the user data and token are removed from localStorage, and the user becomes anonymous (+10);

- **Electronic textbook** (+80):
  -   - [x] the electronic textbook is generated on the basis of a collection of initial data and consists of six sections, each section has 30 pages, each page has 20 words for study (+10);
  -   - [x] the seventh section of the textbook - "Difficult words" is initially empty. This section consists of words that the user has marked as difficult. (+10);
  -  each page of the textbook displays:
     -   - [x] menu (+10);
     -   - [x] list of 20 words (+10);
     -   - [x] links to mini-games "Audio Challenge" and "Sprint" (+10);
     -   - [x] navigation through the pages of the textbook (+10);
     -   - [x] navigation through the seven sections of the textbook with slight differences in the design of each section (+10);
     -   - [x] when reloading the page, the last opened page of the application opens (+10);

- **Word List** (+80):
  -  for each word is displayed:
     -   - [x] the word itself, transcription, translation (+5);
     -   - [x] sentence explaining the meaning of the word, translation of the sentence (+5);
     -   - [x] sentence with an example of the use of the studied word, translation of the sentence (+5);
     -   - [x] picture-association to the studied word (+5);
     -   - [x] audio icon, when clicked, the pronunciation of the word being studied, the pronunciation of the sentence with an explanation of its meaning, the pronunciation of the sentence with an example of its use (+10);
  -  only authorized users are shown:
     -   - [x] a button that, when clicked, can mark a word as a difficult word (in the "Difficult words" section, a button is displayed in its place that unmarks that this is a difficult word and removes it from this section) (+10);
     -   - [x] button, when clicked, the word can be marked as learned (+10);
     -   - [x] if the word is marked as difficult, it remains on the textbook page and is highlighted with a style indicating that this word belongs to difficult words. Also, this word is added to the section "Difficult words" (+10);
     -   - [x] if the word is marked as learned, it remains on the textbook page and is highlighted with a style indicating that this word belongs to the learned words (+10);
     -   - [x] if all the words on a page are learned words or difficult words, that page is considered fully learned and is highlighted by the style. Links to mini-games on this page become inactive. (+10);

- **Mini-games "Audio Challenge" and "Sprint"** (+200):
  - Mini-game "Audio Challenge" (+100):
     -   - [x] the functionality of the mini-game completely repeats the functionality of the mini-game of the same name in the Lingualeo application (+10);
     -   - [x] unified design and UI of mini-game (+10);
     -   - [x] at the end of the game the results of the mini-game are displayed (+10);
     -   - [x] you can control the game with both the mouse and the keys on the keyboard (+10);
     -   - [x] if the mini-game is launched from the menu, one of six difficulty levels can be selected in it (+10);
     -   - [x] if a mini-game is launched from the textbook , it uses the words from the textbook page that contains the link to the game (+10);
     -   - [x] if there are less than 10 words, there is a notification that there are not enough words for the game (+10);
     -   - [x] a series of correct answers is recorded (+10);
     -   - [x] implemented full screen function (+10);
     -   - [x] correct answers are recorded in statistics (+10);

  - Mini-game "Sprint" (+100):
     -   - [x] the functionality of the mini-game completely repeats the functionality of the mini-game of the same name in the Lingualeo application (+10);
     -   - [x] unified design and UI of mini-game (+10);
     -   - [x] at the end of the game, the results of the mini-game are displayed (+10);
     -   - [x] you can control the game with both the mouse and the keys on the keyboard (+10);
     -   - [x] if the mini-game is launched from the menu, one of six difficulty levels can be selected in it (+10);
     -   - [x] a series of correct answers is recorded (+10);
     -   - [x] if there are less than 10 words, there is a notification that there are not enough words for the game (+10);
     -   - [x] points are awarded for correct answers, three correct answers allow user to earn additional points. (+10);
     -   - [x] timer is used (+10);
     -   - [x] implemented full screen function (+10);

- **Study progress** (+50):
  -   - [x] new words are words that were first used in mini-games, regardless of whether the mini-games were opened from the textbook page or from the link in the menu (+10);
  -   - [x] new words are recorded in statistics (+10);
  -   - [x] a series of correct answers is recorded (+10);
  -   - [x] study progress is displayed in statistic (+10);
  -   - [x] progress is available only for authorized users (+10);

- **Learned words** (+60):
  -   - [x] learned words are the words that the user marked as learned on the textbook page (+10);
  -   - [x] if a difficult word has become learned, it is not difficult any more and is removed from the section "Difficult words" (+10);
  -   - [x] learned words are highlighted in style (+10);
  -   - [x] if the learned word has become difficult, it is not learned and is added to the "Difficult words" section (+10);
  -   - [x] difficult words are highlighted in style (+10);
  -   - [x] learned words are not used in mini-games that are launched on the pages of the textbook, but are used in mini-games that are opened via a link in the menu (+10);

- **Statistics page** (+60):
  -   - [x] the statistics page displays short-term statistics for mini-games and words for each day of study (+10);
  - short-term statistics for mini-games indicate the results:
     -   - [x] number of new words per day (+10);
     -   - [x] percentage of correct answers (+10);
     -   - [x] longest series of correct answers (+10);
     -   - [x] number of words learned per day (+10);
     -   - [x] percentage of correct answers per day (+10);

- **Additional functionality** (+40):
  -   - [x] adaptive design (+10);
  -   - [x] high quality application design (+10);
  -   - [x] creative design idea (+10);
  -   - [x] work with svg and animation (+10);
  -   - [ ] long-term statistics for the entire study period not implemented (+0)

### CODE REVIEW

File: src/components/Book/WordItem/WordItem.tsx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  const addActiveWord = (setState: SetState<WordData[]>) => (
+    setState((prev) => [...prev, item])
+  );
+
+  const removeActiveWord = (setState: SetState<WordData[]>) => (
+    setState((prev) => prev.filter((el) => el.id !== item.id))
+  );
+
+  const handleClick = (arr: WordData[], setState: SetState<WordData[]>) => (
+    toggleActive(arr)
```
[the remark this answers is not in the document]
Comment (pug000) [reply src/components/Book/WordItem/WordItem.tsx]: We'll fix it in version 2.0

File: src/components/Book/WordItem/WordItem.tsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            toggleActive(learnedWords)
+              ? defaultTheme.colors.primaryColor
+              : defaultTheme.colors.grey
+          }
+          onClick={() => handleClick(learnedWords, setLearnedWords)}
+        />
+        {
+          toggleActive(difficultWords)
+            ? (
+              <DifficultWordBtnActive
```
[the remark this answers is not in the document]
Comment (pug000) [reply src/components/Book/WordItem/WordItem.tsx]: We'll fix it in version 2.0

File: src/components/DifficultWords/DifficultWordItem/DifficultWordItem.tsx [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  WordTitle
+} from './DifficultWordItem.style';
+
+interface WordItemProps {
+  item: WordData,
+  audio: HTMLAudioElement | null,
+  setNewAudio: (value: HTMLAudioElement | null) => void,
+}
+
+function DifficultWordItem(
```
[the remark this answers is not in the document]
Comment (pug000) [reply src/components/DifficultWords/DifficultWordItem/DifficultWordItem.tsx:48]: We'll fix it in version 2.0

File: src/components/DifficultWords/DifficultWords.tsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  return (
+    <DifficultWordsWrapper>
+      {isLoggedIn
+        && (
+          <>
+            <DifficultWordsTitle>
+              Сложные слова
+            </DifficultWordsTitle>
+            {!wordsDifficult.length && !isLoadingPage
```
[the remark this answers is not in the document]
Comment (pug000) [reply src/components/DifficultWords/DifficultWords.tsx]: We'll fix it in version 2.0

File: src/components/Book/Book.tsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  const [isLoadingPage, setIsLoadingPage] = useState(false);
+
+  useEffect(() => {
+    audio?.remove();
+    setAudio(null);
+
+    (async () => {
+      setIsLoadingPage(true);
+      const data = await getWords(groupNumber, currentPage);
+      setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (pug000) [reply src/components/Book/Book.tsx:89]: We'll fix it in version 2.0

**Review by saachko — APPROVED** [review]

**Review by aArt13 — APPROVED** [review]

**Pull request merged**